astropy/time/tests/test_custom_formats.py

```diff
@@ -127,3 +135,48 @@ def value(self):
 
     finally:
         Time.FORMATS.pop("sort", None)
+
+
+def test_mjd_longdouble_preserves_precision(custom_format_name):
+    class CustomMJD(TimeFormat):
+        name = custom_format_name
+
+        def _check_val_type(self, val, val2):
+            val = np.longdouble(val)
+            if val2 is not None:
+                raise ValueError("Only one value permitted")
+            return val, 0
+
+        def set_jds(self, val, val2):
+            mjd1 = np.float64(np.floor(val))
+            mjd2 = np.float64(val - mjd1)
+            self.jd1, self.jd2 = day_frac(mjd1 + DJM0, mjd2)
+
+        @property
+        def value(self):
+            mjd1, mjd2 = day_frac(self.jd1 - DJM0, self.jd2)
+            return np.longdouble(mjd1) + np.longdouble(mjd2)
+
+    m = 58000.0
+    t = Time(m, format=custom_format_name)
+    t2 = Time(m + 2 * m * np.finfo(np.longdouble).eps, format=custom_format_name)
+    assert t != t2
+    assert isinstance(getattr(t, custom_format_name), np.longdouble)
+    assert getattr(t, custom_format_name) != getattr(t2, custom_format_name)
+
+
+def test_mjd_unit_validation():
+    with pytest.raises(u.UnitConversionError):
+        Time(58000 * u.m, format="mjd")
+
+
+def test_mjd_unit_conversion():
+    assert Time(58000 * u.day, format="mjd") == Time(58000 * u.day, format="mjd")
+    assert Time(58000 * u.day, format="mjd") != Time(58000 * u.s, format="mjd")
+    assert Time(58000 * u.day, format="mjd") == Time(58000 * 86400 * u.s, format="mjd")
+
+
+@pytest.mark.parametrize("f", ["mjd", "unix", "cxcsec"])
+def test_existing_types_refuse_longdoubles(f):
+    with pytest.raises(ValueError):
+        Time(np.longdouble(58000), format=f)
```

Comment:
Not sure we'd want these tests to fail!

Reply:
Ah. Well. Hmm. I want precision loss to require explicit conversion, and none of these types preserve long double precision (yet?).

Reply:
I think all of them start with `day_frac` (in `TimeFromEpoch.set_jds`) so they might all support it if that routine is adjusted.

Reply:
The tests now all permit either outright rejection of long doubles or acceptance provided that the precision is preserved.

Reply:
I'm not sure about precision loss requiring explicit conversion (else raising an unavoidable exception).  This will outright break existing code that happens to be using long doubles, and I'm almost always against breaking code without a strong reason.

In general I prefer to be a bit pragmatic about things like this, letting users be responsible for making decisions about data types, and when it matters then requiring that they have some cognizance of the implementation details.  The second paragraph of the Time documentation starts with "All time manipulations and arithmetic operations are done internally using two 64-bit floats to represent time."  I think that is a pretty good starting point for someone to infer that providing a higher-precision input might not be helping.

But as a compromise it might be fine to issue a warning in the case of potential precision loss.

Reply:
Ideally we just preserve precision in all.... If that isn't currently the case, I'm in favour of warning and setting a separate issue (I think my #9361 will solve it regardless)

Reply:
I'm a bit of an outsider to this conversation, but if there is disagreement on here it seems like something here a kwarg could be used to implement a 'strict' mode that fails if there is precision loss?

Reply:
Long doubles are substantially lower, have to be explicitly selected by the user, and use 50-100% more memory. If someone passes a long double into an astropy function, it is because they want extended precision. 

The astropy Time docs say "All time manipulations and arithmetic operations are done internally using two 64-bit floats to represent time." This says to me, correctly, that the Time object was constructed to make it possible to operate with extended precision. The fact that the *internal* format is based on doubles is no indication at all about how it handles its input. In fact I personally was astonished and horrified to find out that astropy silently discarded my extended precision information. Fortunately I have learned to distrust library writers and test these things.

As another example, PINT has an MCMC based fitting procedure based on the emcee package. In emcee 3.0 a numpy.linalg function was used that raises an exception when presented with long doubles, so I filed a bug report and started pursuing fixes. What I found was that prior to 3.0, emcee was silently discarding extended precision: our MCMC fitter *never* worked, but it failed in a subtle way that we were not aware of. So *please* if you can't provide extended precision, raise an exception when handed an extended-precision data type.

An explicit option to enable strict checking is missing the point: the point is that we should not surprise the user with subtle degradation of their data by default. If users want non-strict checking they can just use np.float on their arguments. The question is not how to support both behaviours, the question is how to avoid surprising our users, and the issue is made more important by the subtlety of the effect that we would be surprising them with.

Reply:
I still think that a Warning in these cases instead of an Exception seems sufficient.  Will that not prevent users from having an expectation of precision that is not realized?

Reply:
I quite strongly feel that we should simply accept higher precision and test that we do it right as much as possible. I don't think, though, that we necessarily have to provide a way to *output* in every format.

Reply:
Agreed.  The warning/exception question is predicated on the current behavior of not necessarily handling higher precision properly.

Reply:
Our code currently emits a blizzard of warnings even for perfectly normal operation, so in fact people grow used to ignoring warnings. This is seen with LaTeX as well - too many warnings and they are ignored.

What is the use case where users pass long doubles in to these functions and would prefer lost precision to exceptions? Recall that manual downconversion is as easy as applying float.

Reply:
If we are not fixing the problem here, then raising an exception seems reasonable.

Reply:
I think of exceptions for issues that are *definitely* wrong or an error, and warnings for things that *might* lead to unexpected results.  I don't think that saying that some code or package issues a blizzard of warnings is a good argument for ignoring the entire Python warning system.

If I do `Time(np.longdouble(1), format='cxcsec')`, there is no reason that should fail to execute.  It is perfectly legal and gives the result that I expect.  But I do think it is reasonable to issue a warning that about precision which an informed user can take into consideration.

It may be that somebody is already using long-double for some reason that we don't know about, and they understand what is happening, and they don't want their code to suddenly break for no good reason with astropy 4.0.  I sometimes do things that might seem irrational or downright stupid to someone else, but we all have our reasons.

Reply:
Ah, convinced the other way again. Let's just agree to fix it!

Reply:
I agree on this - let's get something like #9361 merged and then there won't be much question left what to do with longdoubles because they will work.